src/relations/redis.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Define the Superset server Postgresql relation."""
+
+import logging
+
+from charms.redis_k8s.v0.redis import RedisRelationCharmEvents
+from ops import framework
+from ops.framework import StoredState
+
+from log import log_event_handler
+
+logger = logging.getLogger(__name__)
+
+
+class Redis(framework.Object):
+    """Client for superset:redis relation.
+
+    Attrs:
+        on: redis relation events from redis_k8s library
+        _stored: charm stored state
+    """
+
+    on = RedisRelationCharmEvents()
+    _stored = StoredState()
+
+    def __init__(self, charm):
+        """Construct.
+
+        Args:
+            charm: The charm to attach the hooks to.
+        """
+        super().__init__(charm, "redis")
+        self.charm = charm
+        self.framework.observe(
+            charm.on.redis_relation_updated, self._on_redis_relation_changed
+        )
+
+    @log_event_handler(logger)
+    def _on_redis_relation_changed(self, event):
+        """Handle redis relation updated event.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if self.charm._stored.redis_relation:
+            host, port = self._get_redis_relation_data()
+            self.charm._state.redis_relation = True
+        else:
+            host = None
+            port = None
+            self.charm._state.redis_relation = False
+
+        self.charm._state.redis_host = host
+        self.charm._state.redis_port = port
+
+        self.charm._update(event)
+
+    def _get_redis_relation_data(self):
+        """Get the hostname and port from the redis relation data.
+
+        Returns:
+            redis_hostname: hostname of redis service
+            redis_port: port of redis service
+        """
+        for redis_unit in self.charm._stored.redis_relation:
+            redis_unit_data = self.charm._stored.redis_relation[redis_unit]
+            redis_hostname = redis_unit_data["hostname"]
+            redis_port = redis_unit_data["port"]
+
+        return redis_hostname, redis_port
```

Comment:
If  `self.charm._stored.redis_relation` is empty you are going to get a `NameError` here. But you check that before you use this so it is fine for now. I was going to say that instead of the `if-else` block on line#50, you can have a single call to this function and return a triple from here for `host, port, bool`.

Also, you can iterate with `for redis_unit_data in self.charm._stored.redis_relation.values()` to save a line and an access.

Finally, why do we loop here? Can there be multiple `redis_unit` in the dictionary? Is it safe to override the values of the `host, port` through the loop?

Reply:
There is nothing else in `redis_unit`, the `for` loop was because the key for this dictionary is a relation id that is not accessible via the event. I've found a neater way of doing this now if you could take a second look? It also incorporates the `host, port, bool` tuple returned.